wd_mempool.c

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <poll.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include "wd.h"
+
+struct blkpool {
+	void **blk_elem; /* all the block unit addrs saved in blk_elem */
+	size_t depth;    /* the block pool deph, stack depth */
+	size_t top;      /* the stack top pos for blk_elem */
+	size_t blk_size; /* the size of one block */
+	handle_t mp;     /* record from which mempool */
+};
+
+struct mempool {
+	void *addr;
+	size_t size;
+};
+
+void *wd_blockpool_alloc(handle_t blockpool)
+{
+	struct blkpool *bp = (struct blkpool*)blockpool;
+	
+	if (!bp)
+		return NULL;
+
+	if (bp->top > 0) {
+		bp->top--;
```

Comment:
top is either increased by 1 or decreased by 1. The behavior of top is much like an index. So we could apply "int" type on the "top" field.

Same on "depth" field.